### PR TITLE
Master expense refuse by line yti

### PR DESCRIPTION
**Pull request by tivisse**
TASK: https://www.odoo.com/web#id=28472&view_type=form&model=project.task&action=333&active_id=248&menu_id=4720

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:
## 

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_expense/__manifest__.py [hunk 1]
```diff
@@ -32,8 +32,9 @@
         'security/ir.model.access.csv',
         'data/hr_expense_data.xml',
         'data/hr_expense_sequence.xml',
-        'wizard/hr_expense_refuse_reason.xml',
-        'wizard/hr_expense_register_payment.xml',
+        'wizard/hr_expense_sheet_refuse_reason.xml',
```
Comment (tde-banana-odoo) [inline addons/hr_expense/__manifest__.py]: This is a no.

File: addons/hr_expense/__manifest__.py [hunk 2]
```diff
@@ -32,8 +32,9 @@
...
+        'wizard/hr_expense_refuse_reason_views.xml',
```
Comment (tde-banana-odoo) [inline addons/hr_expense/__manifest__.py]: This is a maybe.

Joke. This is a no.

File: addons/hr_expense/__manifest__.py [hunk 3]
```diff
@@ -32,8 +32,9 @@
...
+        'wizard/hr_expense_sheet_register_payment.xml',
```
Comment (tde-banana-odoo) [inline addons/hr_expense/__manifest__.py]: Xmo is great.

File: addons/hr_expense/models/hr_expense.py [hunk 4]
```diff
@@ -275,6 +281,15 @@ def action_get_attachment_view(self):
         res['context'] = {'default_res_model': 'hr.expense', 'default_res_id': self.id}
         return res
 
+    @api.multi
+    def refuse_expense(self, reason):
+        self.ensure_one()
```
Comment (xmo-odoo) [inline addons/hr_expense/models/hr_expense.py]: y u do this

File: addons/hr_expense/models/hr_expense.py [hunk 5]
```diff
@@ -275,6 +281,15 @@ def action_get_attachment_view(self):
...
+        self.write({'is_refused': True})
```
Comment (xmo-odoo) [inline addons/hr_expense/models/hr_expense.py]: y u do this also

File: addons/hr_expense/models/hr_expense.py [hunk 6]
```diff
@@ -275,6 +281,15 @@ def action_get_attachment_view(self):
         res['context'] = {'default_res_model': 'hr.expense', 'default_res_id': self.id}
         return res
 
+    @api.multi
+    def refuse_expense(self, reason):
```
Comment (xmo-odoo) [inline addons/hr_expense/models/hr_expense.py]: Is there any reason for this to be called directly in RPC?

File: addons/hr_expense/models/hr_expense.py [hunk 7]
```diff
@@ -275,6 +281,15 @@ def action_get_attachment_view(self):
...
+        self.ensure_one()
+        self.write({'is_refused': True})
+        self.sheet_id.write({'state': 'cancel'})
+        body = (_("Your Expense %s has been refused.<br/><ul class=o_timeline_tracking_value_list><li>Reason<span> : </span><span class=o_timeline_tracking_value>%s</span></li></ul>") % (self.name, reason))
```
Comment (xmo-odoo) [inline addons/hr_expense/models/hr_expense.py]: Why wrap everything in parens? Also maybe htmlescape the injected values?

File: addons/hr_expense/wizard/hr_expense_sheet_refuse_reason.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    _description = "Hr Expense Sheet refuse Reason wizard"
+
+    description = fields.Char(string='Reason', required=True)
+
+    @api.multi
+    def expense_refuse_reason(self):
+        self.ensure_one()
+
+        context = dict(self._context or {})
+        active_ids = context.get('active_ids', [])
```
Comment (xmo-odoo) [inline addons/hr_expense/wizard/hr_expense_sheet_refuse_reason.py]: What kind of garbage is that?

File: addons/hr_expense/models/hr_expense.py [hunk 9]
```diff
@@ -43,6 +43,12 @@ class HrExpense(models.Model):
         help="Status of the expense.")
     sheet_id = fields.Many2one('hr.expense.sheet', string="Expense Report", readonly=True, copy=False)
     reference = fields.Char(string="Bill Reference")
+    is_refused = fields.Boolean(string="Explicitely Refused by manager or acccountant", readonly=True, copy=False)
+    refused_reason = fields.Boolean(string="Refused reason", help='''Reason why the line has been explicitely refused by the manager or the accountant:
```
Comment (tde-banana-odoo) [inline addons/hr_expense/models/hr_expense.py]: Why is the reason a boolean ?

By the way, refuse_reason or refused_reason ?

File: addons/hr_expense/models/hr_expense.py [hunk 10]
```diff
@@ -43,6 +43,12 @@ class HrExpense(models.Model):
...
+
+        Example:
```
Comment (tde-banana-odoo) [inline addons/hr_expense/models/hr_expense.py]: I am not sure trivial examples belong to the help parameter.

File: addons/hr_expense/models/hr_expense.py (hunk 6 above)
Comment (tde-banana-odoo) [inline addons/hr_expense/models/hr_expense.py]: Actually we move to action_XXX naming convention when having a workflow-like behavior. So I would rename it action_refuse .By the way, why is it ensure_one, and why don't you store the reason in refuse_reason ?

File: addons/hr_expense/models/hr_expense.py (hunk 7 above)
Comment (tde-banana-odoo) [inline addons/hr_expense/models/hr_expense.py]: I think you could define a small qweb template. We try to remove most of inline html and use templates instead. Easier to customize.

File: addons/hr_expense/models/hr_expense.py [hunk 11]
```diff
@@ -275,6 +281,15 @@ def action_get_attachment_view(self):
         res['context'] = {'default_res_model': 'hr.expense', 'default_res_id': self.id}
         return res
 
+    @api.multi
+    def refuse_expense(self, reason):
+        self.ensure_one()
+        self.write({'is_refused': True})
+        self.sheet_id.write({'state': 'cancel'})
```
Comment (tde-banana-odoo) [inline addons/hr_expense/models/hr_expense.py]: Should'n't we call refuse_expense on the sheet ?

File: addons/hr_expense/wizard/hr_expense_sheet_refuse_reason.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    _name = "hr.expense.sheet.refuse.wizard"
+    _description = "Hr Expense Sheet refuse Reason wizard"
+
+    description = fields.Char(string='Reason', required=True)
+
+    @api.multi
+    def expense_refuse_reason(self):
+        self.ensure_one()
+
+        context = dict(self._context or {})
```
Comment (tde-banana-odoo) [inline addons/hr_expense/wizard/hr_expense_sheet_refuse_reason.py]: Code to clean.

File: addons/hr_expense/models/hr_expense.py (hunk 7 above)
Comment (tde-banana-odoo) [inline addons/hr_expense/models/hr_expense.py]: Check message_post_with_view, for example in forum, using a qweb template.

File: addons/hr_expense/models/hr_expense.py [hunk 13]
```diff
@@ -44,6 +44,12 @@ class HrExpense(models.Model):
         help="Status of the expense.")
     sheet_id = fields.Many2one('hr.expense.sheet', string="Expense Report", readonly=True, copy=False)
     reference = fields.Char(string="Bill Reference")
+    is_refused = fields.Boolean(string="Explicitely Refused by manager or acccountant", readonly=True, copy=False)
+    #refused_reason = fields.Boolean(string="Refused reason", help='''Reason why the line has been explicitely refused by the manager or the accountant:
```
Comment (tivisse) [inline addons/hr_expense/models/hr_expense.py]: Commented code

**Pull request merged**